WordPress/gutenberg
# Block Library: Avoid setting Embed className when empty, non-responsive

This pull request seeks to resolve an issue where embed blocks apply changes immediately upon page load, prompting a user when attempting to navigate away. This only affects themes which do not opt in to [support for responsive embeds](https://wordpress.org/gutenberg/handbook/designers-developers/developers/themes/theme-support/#responsive-embedded-content).

**Implementation notes:**

The root cause is that the embed block assigns a different, new value for the `className` attribute. There is no default value for the `className` attribute, which is reasonable in that we don't want to be assigning an empty `class=""` in markup. However, the internal logic of the embed block in producing class names via `getClassNames` will wrongly produce a different value (`""`) when no changes are necessary (if the class name had been `undefined`).

**Testing instructions:**

These instructions can be repeated on the master branch serving as "Steps to Reproduce" the original issue.

1. (Prerequisite) Activate a theme which does not add theme support for responsive embeds (e.g. "[Astra](https://wordpress.org/themes/astra/)").
2. Navigate to Posts > Add New
3. Click the writing prompt
4. Paste an embeddable URL
   - Example: https://twitter.com/CincinnatiZoo/status/1050770683536568322
5. Save the post
6. Reload the page
7. Verify that (a) the "Save Draft" button is not present but instead "Saved" is displayed and (b) reloading the page does not prompt about unsaved changes

Ensure unit tests pass:

```
npm run test-unit packages/block-library/src/embed/test/util.js
```

Ensure end-to-end tests pass:

```
npm run test-e2e packages/e2e-tests/specs/embedding.test.js
```
